ipa-core/src/query/processor.rs

```diff
@@ -530,57 +698,83 @@ mod tests {
             }
         }
 
         #[tokio::test]
         async fn happy_case() {
-            let network = InMemoryMpcNetwork::default();
-            let identities = HelperIdentity::make_three();
-            let req = prepare_query(identities);
-            let transport = network.transport(identities[1]);
-            let processor = Processor::default();
-
+            let req = prepare_query();
+            let t = TestComponents::new(TestComponentsArgs::default());
             assert!(matches!(
-                processor.query_status(QueryId).unwrap_err(),
+                t.processor.query_status(QueryId).unwrap_err(),
                 QueryStatusError::NoSuchQuery(_)
             ));
-            processor.prepare(&transport, req).unwrap();
+            t.processor
+                .prepare_helper(t.second_transport, t.shard_transport, req)
+                .await
+                .unwrap();
             assert_eq!(
                 QueryStatus::AwaitingInputs,
-                processor.query_status(QueryId).unwrap()
+                t.processor.query_status(QueryId).unwrap()
             );
         }
 
         #[tokio::test]
         async fn rejects_if_coordinator() {
-            let network = InMemoryMpcNetwork::default();
-            let identities = HelperIdentity::make_three();
-            let req = prepare_query(identities);
-            let transport = network.transport(identities[0]);
-            let processor = Processor::default();
+            let req = prepare_query();
+            let t = TestComponents::new(TestComponentsArgs::default());
+            assert!(matches!(
+                t.processor
+                    .prepare_helper(t.first_transport, t.shard_transport, req)
+                    .await,
+                Err(PrepareQueryError::WrongTarget)
+            ));
+        }
 
+        #[tokio::test]
+        async fn rejects_if_not_shard_leader() {
+            let req = prepare_query();
+            let t = TestComponents::new(TestComponentsArgs::default());
             assert!(matches!(
-                processor.prepare(&transport, req),
+                t.processor
+                    .prepare_helper(
+                        t.second_transport,
+                        t.shard_network
+                            .transport(HelperIdentity::TWO, ShardIndex::from(1)),
+                        req
+                    )
+                    .await,
                 Err(PrepareQueryError::WrongTarget)
             ));
         }
 
         #[tokio::test]
         async fn rejects_if_query_exists() {
-            let network = InMemoryMpcNetwork::default();
-            let identities = HelperIdentity::make_three();
-            let req = prepare_query(identities);
-            let transport = network.transport(identities[1]);
-            let processor = Processor::default();
-            processor.prepare(&transport, req.clone()).unwrap();
+            let req = prepare_query();
```

Comment:
can you write a test that ensures if any follower shard on H1 rejects the request, we end up in a clean state on all helpers and all shards?